src/main/java/de/unijena/cheminf/fragment/fingerprint/FragmentFingerprinter.java

```diff
@@ -179,45 +133,50 @@ public IBitFingerprint getBitFingerprint(List<String> aListOfUniqueSmiles) throw
+    public ICountFingerprint getCountFingerprint(Map<String, Integer> aSmilesToFrequencyMap) throws NullPointerException, IllegalArgumentException {
+        Objects.requireNonNull(aSmilesToFrequencyMap, "Given map of string and integer instances is null.");
+        HashMap<Integer, Integer> tmpPositionToFrequencyMap = new HashMap<>(
+                (int) (this.uniqueSmilesToPositionMap.size() * 1.5f),
+                0.75f);
+        String[] tmpSmilesPositionArray = new String[this.uniqueSmilesToPositionMap.size()];
+        for (Map.Entry<String, Integer> tmpEntry : aSmilesToFrequencyMap.entrySet()) {
+            if (tmpEntry.getKey() == null || tmpEntry.getValue() == null) {
+                throw new NullPointerException("Given map of string and integer instances contains " +
                         "instances that are null.");
+            } else if (tmpEntry.getKey().isEmpty() || tmpEntry.getKey().isBlank()) {
+                throw new IllegalArgumentException("Given map of strings an integer instances contains strings that are blank/empty.");
+            } else if (this.uniqueSmilesToPositionMap.containsKey(tmpEntry.getKey())) {
+                //easier debugging (speaking from experience)
+                int tmpPosition = this.uniqueSmilesToPositionMap.get(tmpEntry.getKey());
+                int tmpFrequency = tmpEntry.getValue();
+                tmpPositionToFrequencyMap.put(tmpPosition, tmpFrequency);
             }
-            if(tmpUniqueSmiles.isBlank() || tmpUniqueSmiles.isEmpty()) {
-                throw new IllegalArgumentException("aUniqueSmilesToFrequencyMap (Map of strings an integer instances) contains strings that are blank/empty.");
-            }
-            if (this.uniqueSmilesToPositionMap.containsKey(tmpUniqueSmiles)) {
-                int tmpPosition = this.uniqueSmilesToPositionMap.get(tmpUniqueSmiles);
-                this.cacheRawCountMap.put(tmpPosition,aUniqueSmilesToFrequencyMap.get(tmpUniqueSmiles));
-            }
-            for(int i = 1; i<=aUniqueSmilesToFrequencyMap.get(tmpUniqueSmiles); i++) {
-                this.cacheListToGenerateCountFingerprint.add(tmpUniqueSmiles);
-            }
         }
-        return new CountFingerprint(this.fragmentArray, this.cacheRawCountMap);
+        //uniqueSmilesToPositionMap.keySet().toArray(new String[0]) produces an array where 'map position value != array position'
+        //example: key -> "CCCCC", value -> '26' (position in fingerprint)
+        // in array: because keySet is iterated through as is: 'position' 5 associated with "CCCCC" -> array[5] = "CCCCC"
+        for (Map.Entry<String, Integer> tmpEntry : this.uniqueSmilesToPositionMap.entrySet()) {
+            tmpSmilesPositionArray[tmpEntry.getValue()] = tmpEntry.getKey();
+        }
+        return new CountFingerprint(tmpSmilesPositionArray, tmpPositionToFrequencyMap);
```

Comment:
This is problematic memory-wise if we always create a new array with the bit definitions of the fingerprint for every fingerprint we generate. Is this really necessary? Can we also remove the array from the CountFingerprint class? After a quick scan it doesn't seem like its necessary.